DataViews: Add insert left/right in table column header

ntsekouras opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Resolves: https://github.com/WordPress/gutenberg/issues/72844

This PR adds two menu items in `table` layout that enables users to insert left or right one of the hidden fields.

### Notes
1. I've updated the icons per the mock up [here](https://github.com/WordPress/gutenberg/issues/72844#issuecomment-3468885541) and that this PR doesn't implement any right-click implementation, as it might not be [necessary](https://github.com/WordPress/gutenberg/issues/72844#issuecomment-3480429606).
2. For reviews [hide the whitespace](https://github.com/WordPress/gutenberg/pull/72929/files?diff=split&w=1).




## Testing Instructions
1. Either in site editor or storybook, in a `table` DataViews click the column header and play around with the new menu items (insert left / insert right)
3. Observe that everything works as expected.


## Screenshots or screencast <!-- if applicable -->
<img width="786" height="556" alt="Screenshot 2025-11-06 at 5 52 41 PM" src="https://github.com/user-attachments/assets/5dc0c15d-9efd-4884-bf3d-bee7731d5198" />

ntsekouras committed:
DataViews: Add insert left/right in table column header

oandregal commented on packages/dataviews/src/dataviews-layouts/table/column-header-menu.tsx:
```diff
@@ -187,82 +208,177 @@ const _HeaderMenu = forwardRef( function HeaderMenu< Item >(
+										</Menu.ItemLabel>
+									</Menu.Item>
+								) }
+								{ canMove && !! hiddenFields.length && (
```
```suggestion
								{ !! hiddenFields.length && (
```

ntsekouras replied:
I added the `canMove` check to prevent `insert left` operation. This is the case for primary fields and `view.layout?.enableMoving ?? true` for all other fields. I think it makes sense to keep this check, because even if the `view.layout?.enableMoving` is false, we could still show more fields, that are appended afterwards.

oandregal replied:
Good thinking on the edge cases 👏

We need a more fine-grained approach for this case (a `canInsert` that checks if it's the primary column). Otherwise, this is what happens (only insert right is visible when a table has disabled moving):

<img width="977" height="444" alt="Screenshot 2025-11-05 at 16 19 16" src="https://github.com/user-attachments/assets/8ecfe8f8-9386-4f9b-8814-f87d76165070" />

ntsekouras replied:
>only insert right is visible when a table has disabled moving): 

That's fine though, no? If we add by the view options the field is appended at the end. This is a tiny bit more flexible 😄 

I think it might be enough without a new API for now, but will think about it more. Do you think it's something we should consider now?

oandregal replied:
Yeah, there's no connection between moving things around and adding/removing columns. I don't think this needs to be a new "public" API, just some internal code makes things work like we want.

oandregal replied:
Hey, I've looked at this PR and this is the only thing we need to fix.

When the table layout disables moving:

```js
{
  type: 'table',
  layout: {
    enableMoving: false
  }
}
```

The insert right/left options shouldn't be displayed. What happens now is that the "insert right" still is (screenshot above).

ntsekouras replied:
Cool. I missed `enableMoving`. It seems we need to add a couple of flags to differentiate from `canMove` then, because Primary column sets explicitly `false` and we might want to `insert right`.

ntsekouras replied:
https://github.com/WordPress/gutenberg/pull/72929/commits/40692973ddb778643189984d40a7babcb17a09f0

oandregal commented on packages/dataviews/src/dataviews-layouts/table/column-header-menu.tsx:
```diff
@@ -98,6 +105,20 @@ const _HeaderMenu = forwardRef( function HeaderMenu< Item >(
+		( f ) =>
+			! visibleFieldIds.includes( f.id ) &&
+			! togglableFields.includes( f.id ) &&
+			f.type !== 'media' &&
```
Why do we check for the field type?

ntsekouras replied:
This is the same logic with [view config](https://github.com/WordPress/gutenberg/blob/trunk/packages/dataviews/src/components/dataviews-view-config/index.tsx#L561). If I remember correctly, this to because `type:media` indicates that fields are for choosing alternative media field and are not supposed to render all of them.

oandregal replied:
Can we create an utility that takes the view and the field list as input and returns the hidden fields? This particular logic has grown complex and having it in two places can create inconsistency, so centralizing it would be useful.

ntsekouras replied:
Done here: https://github.com/WordPress/gutenberg/pull/72929/commits/f0f76787778095c8a76aea1fd7e6486c008b680e

ntsekouras committed:
add changelog entry

ntsekouras committed:
remove icons

ntsekouras committed:
add getHiddenFields util

ntsekouras committed:
update util

ntsekouras commented:
I've rebased and updated the util used. I think it's worth it to try this out even with the context menu on table headers. What do you think @jameskoster @youknowriad? In other case, I'll close the PR.

ntsekouras committed:
add canInsertLeft and canInsertRight

oandregal commented:
I've just noticed one thing: we need to disable insert left/right when there's no field hidden.


https://github.com/user-attachments/assets/33e00dbc-8e16-46e1-a323-a03f351d6b2b

ntsekouras committed:
fix bug with empty menu items

ntsekouras commented:
>I've just noticed one thing: we need to disable insert left/right when there's no field hidden.

Good catch. I fixed it.

oandregal commented on packages/dataviews/src/dataviews-layouts/table/column-header-menu.tsx:
```diff
@@ -255,7 +255,7 @@ const _HeaderMenu = forwardRef( function HeaderMenu< Item >(
 								</Menu.Item>
 							) }
-							{ canInsertLeft && (
+							{ canInsertLeft && !! hiddenFields.length && (
```
Can we make it disabled if there are no fields to add?

There's a difference between a menu item that we should never display (e.g., never display insert left in the primary column, or never display left/right if the consumer has disabled moving) vs. a situational context (it's possible to insert fields, but there's none at the moment).

Going to tag @jameskoster for thoughts, but I'm approving in the meantime. It can be addressed as follow-up, if we want to implement what I suggested.

jameskoster replied:
Good question. I don't actually have a strong feeling about this so would welcome input from others. 

My instinct is that after inserting the last option it's kind of implied that there's no further action to take, and therefore reasonable to hide rather than disable. On the other side of the coin, in a DataViews instance where all fields are visible by default, would a disabled 'Insert left' action be confusing?
